### PR TITLE
Prepend `bos token` to Blip generations

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes a failing test reported in our [internal slack](https://huggingface.slack.com/archives/C01NE71C4F7/p1710237799604759?thread_ts=1709544527.619019&cid=C01NE71C4F7). In the prev PR #28994 , we removed `bos` initialization when embeds are passed as input to `generate`, and instead initialized with empty placeholder. As a temporary workaround I suggest to artificially prepend "bos" token after generation. 

For long term, I suggest to reconsider how we work with multimodal LLMs for generation, I guess we'll have more of these models in the future. I analyzed the models we have in more detail, and what I want to do is use the current Llava-style workflow for generation, but with one change. We will need a method similar to "prepare_inputs_for_generation", which will be responsible to taking inputs in all modalities and concatenating them in correct way. The method will be called in generate, while preparing model inputs. I think calling "concat_modalities" from within generate would make it easier for us:

- not to worry about attention and other input lengths mismatch or losing padded attentions, since "concat" will already give us the correct multimodal length. All we do later is simply add one more id for new token
- easy to do assisted decoding with these workflow, which I have been trying to make work for a while 
- we will always have the prompt as part of generated text, which is intuitive for decoder-only models. I guess all VLMs can be thought of as decoder only

I made a board for visual explanation [here](https://miro.com/app/board/uXjVNhBRraU=/?share_link_id=582477932131) and a proof-of-concept that it works [here](https://github.com/zucchini-nlp/transformers/tree/multimodal_generation).

@gante I want your thoughts on this

### CODE REVIEW

File: tests/models/blip_2/test_modeling_blip_2.py [hunk 1]
```diff
@@ -869,7 +869,8 @@ def test_inference_opt(self):
         prompt = "Question: which city is this? Answer:"
         inputs = processor(images=image, text=prompt, return_tensors="pt").to(torch_device, dtype=torch.float16)
 
-        predictions = model.generate(**inputs)
+        # max_length for BLIP includes prompt length, so we need more than 20 to reach EOS here
+        predictions = model.generate(**inputs, max_length=25)
```
Comment (gante) [inline tests/models/blip_2/test_modeling_blip_2.py:873]: Let's use the `max_new_tokens` argument, which prevents prompt length-based issues :)

**Review by gante — COMMENTED** [review]
Regarding the diff in this PR: It can work as a temporary post-generation fix, if we decrement all set length arguments (max_length, max_new_tokens, min_length, min_new_tokens) by 1 to ensure the final length stays consisted with the user's parameterization

**Comment by gante** [comment]
Regarding the generate changes for VLMs: we actually should go beyond that! The key idea is to rewrite generate in a way that it becomes more composable, with a few blocks expected to be overwritten at a model level (like the function you're proposing). Let's chat about that offline, to figure out the requirements :)

File: tests/models/blip_2/test_modeling_blip_2.py (hunk 1 above)
Comment (zucchini-nlp) [reply tests/models/blip_2/test_modeling_blip_2.py]: This uncovered a new problem with BLIP 🙃  Apparently the code was not working properly with `max_new_tokens`.

That was fixed in PR #29585 , so this should be merged only after #29585 is merged and `main` is rebased.

**Review by gante — APPROVED** [review]
LGTM 👍 

(ping me again when this PR's dependency gets merged/approved, so we can merge both)

File: src/transformers/models/instructblip/modeling_instructblip.py [hunk 2]
```diff
@@ -1538,8 +1538,9 @@ def generate(
         inputs_embeds = torch.cat([language_model_inputs, inputs_embeds.to(language_model_inputs.device)], dim=1)
 
         # add image_embeds length to max_length, so that the final max_length in counted only on token embeds
+        # -1 is a temp hack until `generate` is refactored to work better with VLMs
```
Comment (amyeroberts) [inline src/transformers/models/instructblip/modeling_instructblip.py:1541]: @gante's suggestion should also be applied here 

```suggestion
        # -1 is to account for the prepended BOS after `generate.`
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for digging into this and fixing! 

Only request before merging is:
* Run slow model tests
* Make sure there's generation tests which cover this behaviour. Just skimming test_modeling_instructblip.py, it doesn't look like GenerationTesterMixin is used

**Comment by zucchini-nlp** [comment]
@amyeroberts 
1. Yes all blip tests are passing now.
2. Do you mean a test for the BOS token? Maybe we do not really need to check for that, since the PR is actually fixing a failing test. That means there is a test, even though not checking the BOS explicitly.

**Comment by zucchini-nlp** [comment]
@amyeroberts 
1. Yes, only this two models. It is not adding behavior because it was the default behavior for BLIP before a prev PR was merged, which broke it apparently, as we did not run slow tests in VLMs. So this PR fixes it.
2. Oh, I see now. Agreed, I do not know why it was not added. I will take a look if it works with GenerationTesterMixin, otherwise make custom tests.

**Comment by zucchini-nlp** [comment]
I added `GenerationTesterMixin` for blip. But turns out none of the VLMs had tests for generation, because of using both text and pixel values in input, while the Mixin is not ready for it. BLIP is running correctly now, fortunately because BLIP can generate from text only, if image is not passed.

@gante , I think we can update generation tests more, to support model kwargs as input. Then check that all generative models have the GenerationTesterMixin. I can do it in a separate PR 😄

**Comment by gante** [comment]
> @gante , I think we can update generation tests more, to support model kwargs as input. Then check that all generative models have the GenerationTesterMixin. I can do it in a separate PR 😄

👍 Instead of the `GenerationTesterMixin`'s current `_get_input_ids_and_config`, we could have an input preparation function like the `ModelTesterMixin`'s `_prepare_for_class` -- depending on the auto class mapping, prepare the correct inputs for the model, to then feed to `generate`. Looking forward to that rework @zucchini-nlp !

**Review by amyeroberts — APPROVED** [review]
Thanks for working on this fix! 

Multimodal generation tests sound great ❤️  Looking forward to the future PR!

**Pull request merged**